segmenter/video_segmenter_test.go

```diff
@@ -405,3 +408,58 @@ func TestSegmentDefaults(t *testing.T) {
 		t.Errorf("Expected 100ms default segment length but got %v", opt.SegLength)
 	}
 }
+
+func ffprobe_firstframeflags(fname string) (string, error) {
+	cmd := "ffprobe -loglevel quiet -hide_banner -select_streams v -show_packets "
+	cmd = cmd + fname + " | grep flag | head -1"
+	out, err := exec.Command("bash", "-c", cmd).Output()
+	return strings.TrimSpace(string(out)), err
+}
+
+func TestMissingKeyframe(t *testing.T) {
+	// sanity check that test file has a keyframe at the beginning
+	out, err := ffprobe_firstframeflags("test.flv")
+	if err != nil || out != "flags=K_" {
+		t.Errorf("First video packet of test file was not a keyframe '%v' - %v", out, err)
+		return
+	}
+
+	// remove the first keyframe from test file, store in tempfile
+	dir, err := ioutil.TempDir("", "lp-"+t.Name())
+	defer os.RemoveAll(dir)
+	if err != nil {
+		t.Errorf(fmt.Sprintf("Unable to get tempfile ", err))
+		return
+	}
+	fname := path.Join(dir, "tmp.flv")
+	oname := path.Join(dir, "out.m3u8")
+	cmd := "-i test.flv -bsf:v noise=keyframes=1 -c:a copy -c:v copy -copyinkf -y " + fname
+	c := exec.Command("ffmpeg", strings.Split(cmd, " ")...)
+	err = c.Run()
+	if err != nil {
+		t.Errorf(fmt.Sprintf("Unable to run 'ffmpeg %v' - %v", cmd, err))
+		return
+	}
+
+	// sanity check tempfile doesn't have a video keyframe at the beginning
+	out, err = ffprobe_firstframeflags(fname)
+	if err != nil || out != "flags=__" {
+		t.Errorf("First video packet of temp file unexpected; %v - %v", out, err)
+		return
+	}
+
+	// actually segment
+	ffmpeg.InitFFmpeg()
+	defer ffmpeg.DeinitFFmpeg()
+	err = ffmpeg.RTMPToHLS(fname, oname, path.Join(dir, "out")+"_%d.ts", "4")
+	if err != nil {
+		t.Errorf("Error segmenting - %v", err)
+		return
+	}
+	// and now check that segmented result does have keyframe at beginning
+	out, err = ffprobe_firstframeflags(path.Join(dir, "out_0.ts"))
```

Comment:
Just for personal clarification...is the idea that if the input file does not have a video keyframe, the segmenter will make sure that the first segment does have a keyframe?

Reply:
Correct, the segmenter will drop the packets until it gets a keyframe. Truth be told, this isn't a great solution for a few reasons:

* If we don't have video it drops everything. (The same has been true for audio though.)
* It drops audio that may come in before video, or "late" video
* If it doesn't get a video keyframe, it drops everything also. This video is probably un-decodeable anyway, but it may still be useful to write just the audio to disk.

Fixing these the correct way will take a bit more work than this short-term solution.

Related: https://github.com/livepeer/lpms/issues/60

Reply:
Ah ok got it yeah I confused myself previously - I was thinking that the segmenter somehow adds a keyframe when in reality it just drops packets until it gets a keyframe such that the result is that the first segment outputted by the segmenter necessarily contains a keyframe